### PR TITLE
Factor out the `BasicAuthAuthenticationPolicy` from Pyramid 1.x

### PR DESCRIPTION
**Pull request by jon-betts**
For: https://github.com/hypothesis/h/issues/6595

This is one of the actual steps towards Pyramid 2.0 (https://docs.pylonsproject.org/projects/pyramid/en/2.0-branch/whatsnew-2.0.html#upgrading-auth-20)

### In this PR

 * Removes the deprecated `BasicAuthAuthenticationPolicy`
 * This was achieved by inlining all the code, then refactoring it
 * This also inlines a "helper" function which was just private code in another file
 * This also introduces the `identity()` method as required by Pyramid 2.0 security policies, although we only use it for internal purposes for the moment.

### Review notes

 * The commits show some baby refactoring steps along the way, but the last one is a belter
 * As per usual this involved rewriting the tests in question
 * `h.auth.util.verify_auth_client` has now gone and is inline
 * I can't stress enough how little I like the pattern of having loads of classes in one file and loads of private functions for them in another. I _really_ think we should avoid this and just put the code where it's used

### Testing notes

The `TokenAuthenticationPolicy` is used when logging in with a bearer token. This is basically what the client does. So if the client works, this works.

This means fire up the usual set of `h`, LMS and the Via's and noodle about a bit.

### CODE REVIEW

File: h/auth/policy.py [hunk 1]
```diff
@@ -341,6 +197,103 @@ def forget(self, _request):  # pylint: disable=no-self-use
...
+
+    A client can also pass an `X-Forwarded-User` header with the userid to
+    act as user in their authority. This will create a `request.user` and will
+    look like a normal login. This user will have an additional
+    principal, `client:{client_id}@{authority}` which lets you tell it apart
+    from regular users.
+
+    To differentiate between request with a token authenticated user and
+    a request with a forwarded user, the latter has an additional
+    principal, `client:{client_id}@{authority}`.
```
Comment (marcospri) [inline h/auth/policy.py:217]: This might be the same information twice?

File: h/auth/policy.py [hunk 2]
```diff
@@ -341,6 +197,103 @@ def forget(self, _request):  # pylint: disable=no-self-use
...
+            return None
+
+        if (
+            # The client must exist
+            auth_client is None
+            # Have a secret to compare with
+            or auth_client.secret is None
+            # And be the correct type
+            or auth_client.grant_type != GrantType.client_credentials
+        ):
```
Comment (marcospri) [inline h/auth/policy.py:281]: Much nicer than the original ifs chain

**Review by marcospri — APPROVED** [review]
The end result is a nice refactor :+1:

File: h/auth/policy.py (hunk 1 above)
Comment (jon-betts) [reply h/auth/policy.py:217]: Oops yeah. I was writing the one above to try and capture all the information in the previous description but shorter. I missed out deleting this part as I expressed it.

**Pull request merged**